### PR TITLE
Tests for igraph_closeness_estimate

### PR DESCRIPTION
**Pull request by YasirKusay**
I've implemented a few tests for igraph_closeness_estimate for as many cases that I can think of. I will admit that the code may be unnecessarily complicated to read and that some of the tests are too simple. Furthermore, I have yet to implement tests for specific cutoff values, rather than just calculating the true closeness centrality as I am confused as to how it works. From what I understand, if the path from one node to another exceeds the cutoff value, the cutoff value is used instead but the output I am getting is not consistent with this. Feel free to give me advice as how to improve the tests and the code, and if I need to change my code in order to conform to your test protocols.

### CODE REVIEW

**Comment by szhorvat** [comment]
Thanks for all your work on this @YasirKusay !

Since you are still working on it, I think it's best to merge this only *after* the 0.8 release. However, if you discover any bugs while creating the tests, please let us know ASAP.

**Comment by szhorvat** [comment]
In case you have not yet seen this wiki page, here are some tips on what to verify: https://github.com/igraph/igraph/wiki/Checklist-for-new-(and-old)-functions

**Comment by vtraag** [comment]
Yes, indeed thanks for you work @YasirKusay ! As @szhorvat indicates, we should only merge this after the upcoming release.

One other thing: it is useful to indicate in your PR which issue you intend to fix (if any). This will both link the PR to the issue, and indicate in the issue which PR fixes it. In this case, you intend to address issue #1208. If you indicate something like: "this should fix #1208", GitHub even automatically picks up on this and closes the issue if we merge the PR. You only need to write the issue number (prepended with `#`), GitHub will pick up on this and link to the appropriate issue.

**Comment by YasirKusay** [comment]
Thank you all very much for your advice. I will continue to work on it for now. These tests were intended to address issue #1208 (add tests for igraph_closeness). One more thing, can you explain to me how the cutoff parameter is meant to work? From what I understand, if the path from one node to another exceeds the cutoff value, the cutoff value is used instead but the output I am getting is not consistent with this.

**Comment by szhorvat** [comment]
Just a note that I haven't forgotten about this and will get back to you in a couple of days. Thank again for doing all this work!

**Comment by szhorvat** [comment]
Can you please separate changes related to different issues? This PR was started for adding tests for closeness calculations. Things related to bridges should be kept separate.

**Comment by iosonofabio** [comment]
Hi Yasir,

Thank you for your commitment! The usual way to submit several PRs at once is to have separate branches (both locally on your laptop and remote branches on your GitHub) and then each remote branch is used for one PR. Git will take care of merging them all.

**Comment by szhorvat** [comment]
> One more thing, can you explain to me how the cutoff parameter is meant to work? From what I understand, if the path from one node to another exceeds the cutoff value, the cutoff value is used instead but the output I am getting is not consistent with this.

This related to issue https://github.com/igraph/igraph/issues/1053 : igraph's current handling of non-connected graphs is rather questionable in my opinion ... We proceeding, we need to think about this carefully, and choose how to handle disconnected graphs, or cases with look "disconnected" with a low cutoff.

**Comment by YasirKusay** [comment]
With regards to issue #1053, I definitely agree! When I was writing tests for a weighted graph, I found it unusual that I was using the number of vertices as the default distance for vertices in separated subgraphs (or even in connected, directed graphs where there was no possible way to reach a vertex from another vertex, e.g. in 1->2->3, where you cannot reach node 2 from node 3). Furthermore, say we have a small graph: 1-2-3, where the distance from 1-2 is 10 and 2-3 is 20, therefore disproving that the number of vertices is " is always longer than the longest possible geodesic". I actually did a bit of research into this and I found some very interesting ways of working out closeness centrality, particularly this one: 

https://toreopsahl.com/2010/03/20/closeness-centrality-in-networks-with-disconnected-components/

To summarise what that website stated, since the distance between 2 disconnected nodes is infinite, the reciprocal of that distance will be 0 and therefore useless when you are finding the reciprocal of the sum of all distances, as centrality will be 0. Instead, it proposes that you find the distance between 2 nodes and get the reciprocal of that, (doing the same from that node to all of the nodes in the graph, even if they are disconnected) and then sum up these reciprocals. I acknowledge that this will result in a different closeness centrality for graphs that are connected, but perhaps it can be implemented for disconnected graphs and directed graphs where there is no possible path from one node to another (whilst triggering a warning to the user).

**Comment by YasirKusay** [comment]
@szhorvat, @iosonofabio 

Thank you for your advice and I will do that from now on. I am a bit new to github and didnt realise that commits are connected to the same pull request.

**Comment by vtraag** [comment]
@YasirKusay, you may want to read the general [contributing](https://github.com/igraph/igraph/blob/master/CONTRIBUTING.md) guidelines. That also contains some explanations of how to best prepare your PR (it also took me some time to learn it).
